### PR TITLE
Refactor Rake

### PR DESCRIPTION
**Pull request by SamLR**
This refactors the rake tasks & merges those from govuk-dns-config.

It:
* moves the terraform tasks to their own namespace (`tf:`
* moves "utility" tasks (e.g. `validate_generate_environment`) to functions
* makes environment variable usage more consistent across tasks.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,124 +1,182 @@
...
 
-```
-bundle install
-bundle exec rake plan
-```
+* GOOGLE_DNS_NAME - The DNS domain that will be deployed to (e.g. `example.com.`, GCE needs fully qualified domain names for its entries).
+* GOOGLE_ZONE_NAME - name of the hosted zone.
+* GOOGLE_PROJECT - project ID
+* GOOGLE_REGION - project region
+* GOOGLE_CREDENTIALS - JSON credentials for the service account to deploy using (best set using `GOOGLE_CREDENTIALS=(cat <path to credentials>`).
```
Comment (surminus) [inline README.md]: Should `$(cat <path to credentials>)`

File: README.md [hunk 2]
```diff
@@ -1,124 +1,182 @@
...
 
-#### Export the environment applicable to your credentials
-For example:
-```
-export DEPLOY_ENV=test
+#### Dyn ####
 
-export PROVIDERS=route53
-```
+* DYN_ZONE_ID - which zone to deploy to
```
Comment (surminus) [inline README.md]: Can we have these in shell markdown ie `wrapped in backticks` to make it consistent with documentation above?

**Comment by surminus** [comment]
Couple of minor comments on the documentation but the changes look sensible. I haven't tested locally yet though, but I will have a look and approve if it all seems to work as intended.

I'm still not entirely convinced how easy it would be for someone to pick up this repo and use it to deploy with. The documentation is clear, but it should be easier to "pick up and play". Perhaps a "Getting Started in 5 minutes" guide at the top would help, with then further verbose documentation further on might be useful?

File: README.md [hunk 3]
```diff
@@ -4,6 +4,7 @@ A set of rake tasks for managing DNS records using [Terraform](https://terraform
 
 | Table of Contents |
 | ----------------- |
+| [Quick Start](#quick-start) |
 | [The Tasks](#the-tasks) |
```
Comment (surminus) [inline README.md]: just noticed that the 'The Tasks' and 'Installation' are the wrong way round in the table of contents.

**Comment by surminus** [comment]
I'm not sure if it's just me, but I'm getting an error when trying to run the `validate_dns` task now:

```
🐫  ~/govuk/govuk-dns git:(tidy-rake) ✗ ZONEFILE=../govuk-dns-config/publishing.service.gov.uk.yaml brake validate_dns

An error occurred while loading ./spec/validate_published_dns_spec.rb.
Failure/Error: records = resolver.query(query, 'ANY')

Dnsruby::ServFail:
  Dnsruby::ServFail
# /Users/lauramartin/.bundle/ruby/2.3.0/gems/dnsruby-1.60.1/lib/dnsruby/resolver.rb:257:in `send_message'
# /Users/lauramartin/.bundle/ruby/2.3.0/gems/dnsruby-1.60.1/lib/dnsruby/resolver.rb:203:in `query'
# ./spec/validate_published_dns_spec.rb:49:in `block (3 levels) in <top (required)>'
# ./spec/validate_published_dns_spec.rb:40:in `block (2 levels) in <top (required)>'
# ./spec/validate_published_dns_spec.rb:39:in `each'
# ./spec/validate_published_dns_spec.rb:39:in `block in <top (required)>'
# ./spec/validate_published_dns_spec.rb:22:in `<top (required)>'
Run options: include {:validate_dns=>true}

Finished in 0.00038 seconds (files took 3.75 seconds to load)
0 examples, 0 failures, 1 error occurred outside of examples
```

**Comment by SamLR** [comment]
That error is caused by the record not being locatable (for example if the parent domain were missing)

**Comment by SamLR** [comment]
Merge with this [PR](https://github.com/alphagov/govuk-puppet/pull/6011)

**Pull request merged**